### ISSUE TITLE
Not Validating the authority for Absolute Urls

### ISSUE THREAD

**Issue by sampathsree** [lemonlabsuk/scala-uri#98]
AbsoluteUrl parsing is not validating the domain name in authority to only allow LDH ("letter digit hyphen") characters. Where as standard Java URI honors it. This is what it is mentioned in RFC-3986

`
Such a name consists of a sequence of domain labels separated by ".", each domain label starting and ending with an alphanumeric character and possibly also containing "-" characters
`

**To Reproduce**
```scala
import io.lemonlabs.uri.AbsoluteUrl
import java.net.URI

scala> println(AbsoluteUrl.parse("https://www.goog le.com?q=i+am+invalid"))
https://www.goog le.com?q=i%2Bam%2Binvalid

scala> println(new URI("https://www.goog le.com?q=i+am+invalid").toURL)
java.net.URISyntaxException: Illegal character in authority at index 8: https://www.goog le.com?q=i+am+invalid
  at java.net.URI$Parser.fail(URI.java:2848)
  at java.net.URI$Parser.parseAuthority(URI.java:3186)
  at java.net.URI$Parser.parseHierarchical(URI.java:3097)
  at java.net.URI$Parser.parse(URI.java:3053)
  at java.net.URI.<init>(URI.java:588)
  ... 36 elided
```

**Comment by theon** [comment]
Thanks for raising this. I can change the parser to only allow characters in RFC-3986 (see below). That would solve your example. Let me know if that works for you.

```
unreserved  = ALPHA / DIGIT / "-" / "." / "_" / "~"
pct-encoded = "%" HEXDIG HEXDIG
sub-delims  = "!" / "$" / "&" / "'" / "(" / ")" / "*" / "+" / "," / ";" / "="
reg-name    = *( unreserved / pct-encoded / sub-delims )
```

I would like to keep the existing behaviour as well, but would likely put that behind some config. The reason for this is that despite what the spec says, I can see that some HTTP clients and servers are more permissive:

```
$ sudo echo '140.82.118.4 git[h]ub.com' >> /etc/hosts
$ curl -v 'http://git[h]ub.com'
*   Trying 140.82.118.4...
* TCP_NODELAY set
* Connected to git[h]ub.com (140.82.118.4) port 80 (#0)
> GET / HTTP/1.1
> Host: git[h]ub.com
> User-Agent: curl/7.64.1
> Accept: */*
> 
< HTTP/1.1 301 Moved Permanently
< Content-length: 0
< Location: https://git[h]ub.com/
< 
```

But since is likely rare that people are using weird hostnames like this, I'm happy for the stricter parsing (as per the spec) to be the default and some config needed to enable to permissive parsing:

```scala
AbsoluteUrl.parse("https://www.goog le.com?q=i+am+invalid") // Failure
```

```scala
implicit val config: UriConfig = UriConfig(strictRegName = false)
AbsoluteUrl.parse("https://www.goog le.com?q=i+am+invalid") // Success
```

**Comment by sampathsree** [comment]
> Thanks for raising this. I can change the parser to only allow characters in RFC-3986 (see below). That would solve your example. Let me know if that works for you.
> 
> ```
> unreserved  = ALPHA / DIGIT / "-" / "." / "_" / "~"
> pct-encoded = "%" HEXDIG HEXDIG
> sub-delims  = "!" / "$" / "&" / "'" / "(" / ")" / "*" / "+" / "," / ";" / "="
> reg-name    = *( unreserved / pct-encoded / sub-delims )
> ```
> 
> I would like to keep the existing behaviour as well, but would likely put that behind some config. The reason for this is that despite what the spec says, I can see that some HTTP clients and servers are more permissive:
> 
> ```
> $ sudo echo '140.82.118.4 git[h]ub.com' >> /etc/hosts
> $ curl -v 'http://git[h]ub.com'
> *   Trying 140.82.118.4...
> * TCP_NODELAY set
> * Connected to git[h]ub.com (140.82.118.4) port 80 (#0)
> > GET / HTTP/1.1
> > Host: git[h]ub.com
> > User-Agent: curl/7.64.1
> > Accept: */*
> > 
> < HTTP/1.1 301 Moved Permanently
> < Content-length: 0
> < Location: https://git[h]ub.com/
> < 
> ```
> 
> But since is likely rare that people are using weird hostnames like this, I'm happy for the stricter parsing (as per the spec) to be the default and some config needed to enable to permissive parsing:
> 
> ```scala
> AbsoluteUrl.parse("https://www.goog le.com?q=i+am+invalid") // Failure
> ```
> 
> ```scala
> implicit val config: UriConfig = UriConfig(strictRegName = false)
> AbsoluteUrl.parse("https://www.goog le.com?q=i+am+invalid") // Success
> ```

That looks promising, but I am not sure having `strictRegName` is a good idea. `AbsoluteUrl` by definition means it must be a standard URL with a scheme and a valid host. I understand about the need for permissive parsing, may be we should have another generic url type under `UrlWithAuthority` trait. Also for host parsing this is what recommended on WHATWG - https://url.spec.whatwg.org/#host-writing But if you still believe having `strictRegName` is better alternative, at least I would prefer it to be `true` by default for `AbsoluteUrl`. Thanks!

**Comment by theon** [comment]
My reasoning for something like `strictRegName` is that I suspect out in the wild there are some weird hostnames being used that go against both the WHATWG and RFC 3986 specs and I would like to keep support for those. I agree that this shouldn't be the default behaviour but have to be explicitly turned on. I'm not sure a `UrlWithAuthority` subtype would help as `Uri.parse` and `Url.parse` would still need something to know if it should do permissive or strict parsing.

Maybe something like `strictHostParsing` is a better name.

Thinking about your example more, I think I'd be happy to always prevent spaces in hosts (even with `strictHostParsing=false`), since I don't think any HTTP clients/servers would actually support non percent encoded spaces in a host. That said, I have got hosts with other illegal characters to be accepted by HTTP clients and servers which is what I'd like `strictHostParsing=false` to support.

Related: Thanks for the WHATWG link. Whilst I'm not keen on a new `UrlWithAuthority` subtype, there might be a case for a new subtype under `io.lemonlabs.uri.Host`. The WHATWG spec states:
> A host is a domain, an IPv4 address, an IPv6 address, an opaque host, or an empty host.

Where `opaque host` looks to be more permissive than `domain`. Currently scala-uri has the subtypes `DomainName`, `IpV4` and `IpV6`, so maybe adding an `OpaqueHost` makes sense. I need to spend some more time reading the WHATWG spec to work out if that would be something useful.

**Comment by sampathsree** [comment]
`strictHostParsing` on top level (like `Url` or `Uri`) seems to be a good option for supporting permissive parsing. And yes, we should not allow host with spaces in either case. Thanks for supporting the request.

**Comment by theon** [comment]
I started implementing the `strictHostParsing` mentioned above, but then realised that it would break the current [IDN/punycode support](https://github.com/lemonlabsuk/scala-uri#punycode), so I have simply made it so whitespace is not allowed in the host to fix the issue in the original post. 

Note that `java.net.URI` also supports internationalized domains
```scala
scala> new URI("https://はじめよう.みんな/howto.html").toURL
res2: java.net.URL = https://はじめよう.みんな/howto.html
```

**Comment by theon** [comment]
@sampathsree This is now released in version `2.1.0`. Let me know if it solves your issue. Thanks again
